pandas/io/formats/csvs.py

```diff
@@ -69,7 +69,10 @@ def __init__(
     ) -> None:
         self.fmt = formatter
 
-        self.obj = self.fmt.frame
+        if hasattr(self.fmt.frame, "sparse"):
```

Comment:
Does this only catch cases where the entire dataframe is sparse? What about when only some columns are `SparseArray`?

I worry a bit here about the forced materialization of the full sparse data at once. Since formatting occurs column-by-column, would it be possible to instead materialize each _column_ as necessary instead of the entire frame at once? Similarly, this has the downside of materializing the full dataframe without regard for `chunksize`. `chunksize` will always make serialization slower, but not respecting it (along with column issue) might cause problematically higher memory usage.

Reply:
Is it possible to instead handle this later by adjusting `_format_native_types`?

Reply:
Based on my [profiling](https://gist.github.com/rtlee9/bf5cb69494532337f9533598624a7a5d) `_format_native_types` did not account for any of the difference in performance between forced materialization (fast) and the current behavior (slow) so wouldn't help as a result. The same profiler [showed](https://gist.github.com/rtlee9/ca34f6693312c3f5122753b703cf2c2f) that `to_native_types`, which is called earlier in `_save_chunk`, was more of a bottleneck due to number of calls rather than time / call.  Looks like this happens because there are many more blocks without the forced materialization (maybe one per row).


We could potentially implement this logic deeper in the call stack such as in `managers.to_native_types` but I'm not familiar with this class and it seemed like it could have broader-than-intended effects.

Reply:
I'm not sure I follow what you're saying about chunksize. AFAIK this PR doesn't change the chunking behavior in `_save_body` in that `chunksize` is unchanged and there are the same number of calls to `_save_chunk`. I do understand the idea that materializing sparse data can consume more memory but still trying to connect this with chunksize in my mind.

For a dataframe with mixed sparse and dense columns you're right that this PR won't do anything. `hasattr(pd.DataFrame({"A": pd.arrays.SparseArray([0, 1]), "B": [0, 0]}), 'sparse')` evaluates to false. Seems like materializing column-by-column could help with this particular case but is it worth the complexity if that is the only benefit (i.e., is this a common scenario)?

Reply:
> I'm not sure I follow what you're saying about chunksize. AFAIK this PR doesn't change the chunking behavior in `_save_body` in that `chunksize` is unchanged and there are the same number of calls to `_save_chunk`. I do understand the idea that materializing sparse data can consume more memory but still trying to connect this with chunksize in my mind.
> 
The way I thought this worked (could be wrong :) was essentially the following psuedocode:
```
for chunk in chunks
   do_format(chunk)  # Causes sparse to dense conversion
   ...
```
So the main concern is that this pr causes a change from `1 chunksize of data converted to dense at a time` -> `whole data converted to dense at once`. So while an awesome perf improvement, does this cause problems for sparse data that can't be fit into memory?

Reply:
Yes, agreed. We should materialise column by column. This is what I referenced initially

Reply:
Ok so it sounds like there are two concerns:
1. Materializing the entire frame at once rather than in chunks can increase memory consumption drastically
2. This pr does not currently affect frames that contain some sparse columns and some dense columns

For 1) this makes a lot of sense that this could be an issue and 6d092ee6 addresses this by moving the materialization to `_save_chunk`. Unsurprisingly this erases some of the performance gains (see comment below) but I agree this is worth the memory preservation. In the future maybe we can do something dynamic where we optionally convert the whole frame to dense all at once to reap the additional perf gains but only if sufficient memory (or if the user specifies).

For 2) I'm not sure the best way to implement column-wise materialization while preserving the perf gains. A simple iteration over columns and casting to_dense results in the same number of blocks and no perf gains, so would need to do something beyond this.

Reply:
This should be done in ``to_native_types`` I think, this is in a different place now, but has the same issues as before

Reply:
It resolves the first of the two issues I outlined above — it materializes within chunks to save memory